In [150]:
# Initialize Otter
import otter
grader = otter.Notebook("ps6.ipynb")

In [151]:
import numpy as np

## Question 1: Down with `for` loops
Each of the problems below contains a function that uses `for` loops to perform a certain operation on arrays. Your job is to rewrite these function using only Numpy array manipulations and library functions (e.g. `np.xxx()`). Do not use any `for` or `while` loops, iterators, generators, or list comprehensions in your solutions.

**1(a)** (3 pts) Return all of the rows of the integer matrix `A` where where each entry of the row is distinct:

In [152]:
def distinct_rows_py(A):
    'Return all rows of A that have completely distinct entries.'
    return np.array([a for a in A if len(set(a)) == len(a)])

In [153]:
A = np.eye(5)
distinct_rows_py(A)

array([], dtype=float64)

In [154]:
A = np.arange(9).reshape(3, 3)
distinct_rows_py(A)

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [155]:
if all(np.unique(A,return_counts = True, axis = 0)[1]):
    print(1)

1


In [156]:
a = np.unique(A,return_counts = True, return_index = True)
a

(array([0, 1, 2, 3, 4, 5, 6, 7, 8]),
 array([0, 1, 2, 3, 4, 5, 6, 7, 8]),
 array([1, 1, 1, 1, 1, 1, 1, 1, 1]))

In [157]:
A = np.arange(9).reshape(3,3)

In [158]:
A = np.array([
    [1, 2, 3],
    [4, 4, 4],
    [5, 6, 6]])
distinct_rows_py(A)

array([[1, 2, 3]])

In [159]:
A[:,::-1]

array([[3, 2, 1],
       [4, 4, 4],
       [6, 6, 5]])

In [160]:
def row_unique(a):
    unique = np.sort(a)
    duplicates = unique[:,  1:] == unique[:, :-1]
    
    print(~duplicates) # the ~ is the same as the np.invert()
    
    unique[:, 1:][duplicates] = 0
    allRowTrue = np.all(~duplicates ==True , axis = 1)
    print(allRowTrue)
    return unique[allRowTrue]

In [12]:
row_unique(A)

[[ True  True]
 [False False]
 [ True False]]
[ True False False]


array([[1, 2, 3]])

In [13]:
A.shape

(3, 3)

In [14]:
def distinct_rows_np(A):
    repeat = A[:, 1:] != A[:, :-1]
    allRowTrue = np.all(repeat == True, axis = 1)  # check in the horizontal direction
    return A[allRowTrue]

In [15]:
grader.check("1a")

1a results: All test cases passed!

**1(b)** (3 pts) Given a vector $v$ of length $n$, and an integer $0<k<n$, return the matrix
```
[[v[0], ..., v[k-1]],
 [v[1], ..., v[k]  ],
 [v[2], ..., v[k+1]],
 [ .     .       . ],
 [ .     .       . ],
 [ .     .       . ],
 [v[n-k+1, ..., v[n]]
 ```

In [16]:
def sliding_stack_py(v, k):
    "Stack sliding windows of v of length k."
    rows = []
    for i in range(len(v) - k + 1):
        rows.append(v[i : (i + k)])
    return np.array(rows)

In [17]:
sliding_stack_py(np.array([1, 2, 3, 4, 5]), 3)

array([[1, 2, 3],
       [2, 3, 4],
       [3, 4, 5]])

In [18]:
x = np.array([1,2,3,4,5])


In [19]:
x[np.array([[1,2,3],[2,3,4]])]

array([[2, 3, 4],
       [3, 4, 5]])

In [20]:
np.arange(6)[None, :]

array([[0, 1, 2, 3, 4, 5]])

In [21]:
np.arange(4)[:, None]

array([[0],
       [1],
       [2],
       [3]])

In [22]:
np.arange(6)[None, :] + np.arange(4)[:, None]

array([[0, 1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5, 6],
       [2, 3, 4, 5, 6, 7],
       [3, 4, 5, 6, 7, 8]])

In [23]:
def sliding_stack_np(v, k):

    indexArray = np.arange(k)[None, :] + np.arange(len(v) - k + 1)[:, None]  
    return v[indexArray]

In [24]:
sliding_stack_np(x,3)

array([[1, 2, 3],
       [2, 3, 4],
       [3, 4, 5]])

In [25]:
grader.check("1b")

1b results: All test cases passed!

**1(c)** (3 pts) Given a vector of non-negative integers `v`, with `max(v) = m`, return a vector `c` of length `m + 1` such that `c[i]` is the number of times that the integer `i` appears in `v`.

In [26]:
def digit_count_py(v):
    m = max(v)
    ret = np.zeros(m + 1, int)
    for vv in v:
        ret[vv] += 1
    return ret

In [27]:
v = np.array([0, 0, 1, 1, 2, 2, 2, 3, 4])
digit_count_py(v)

array([2, 2, 3, 1, 1])

In [28]:
v = np.append(v,np.arange(4))
v

array([0, 0, 1, 1, 2, 2, 2, 3, 4, 0, 1, 2, 3])

In [29]:
np.unique(v,return_counts = True)[1] - 1

array([2, 2, 3, 1, 0])

In [30]:
def digit_count_np(v):
    m = max(v)
    v = np.append(v, np.arange(m + 1))
    return np.unique(v, return_counts = True)[1] - 1

In [31]:
grader.check("1c")

1c results: All test cases passed!

**1(d)** (3 pts) Call a square $n\times n$ matrix $A$ *countersymmetric* if $A_{ij} = A_{n-j,n-i}$ for all $i$ and $j$. An example of such a matrix is:

$$
\begin{pmatrix}
4 & 3 & 2 & 1 & 0\\
8 & 7 & 6 & 5 & 1\\
11 & 10 & 9 & 6 & 2\\
13 & 12 & 10 & 7 & 3\\
14 & 13 & 11 & 8 & 4
\end{pmatrix}
$$

Write a function `is_countersym` that checks this property:

In [32]:
def is_countersym_py(A):
    "Returns True if A is countersymmetric"
    n = A.shape[0]
    for i in range(n):
        for j in range(n):
            if A[i, j] != A[n - j - 1, n - i - 1]:
                return False
    return True

In [33]:
cs_matrix = np.array([[ 4,  3,  2,  1,  0], [ 8,  7,  6,  5,  1], [11, 10,  9,  6,  2], [13, 12, 10,  7,  3], [14, 13, 11,  8,  4]])
is_countersym_py(cs_matrix)

True

In [34]:
cs_matrix = np.array([[ 4,  3,  2,  1,  0], [ 8,  7,  6,  5,  1], [11, 10,  9,  6,  2], [13, 12, 10,  7,  3], [14, 13, 11,  8,  4]])
cs_matrix[0] == np.flip(cs_matrix[:,-1])

array([ True,  True,  True,  True,  True])

In [35]:
cs_matrix

array([[ 4,  3,  2,  1,  0],
       [ 8,  7,  6,  5,  1],
       [11, 10,  9,  6,  2],
       [13, 12, 10,  7,  3],
       [14, 13, 11,  8,  4]])

In [36]:
np.flip(cs_matrix)

array([[ 4,  8, 11, 13, 14],
       [ 3,  7, 10, 12, 13],
       [ 2,  6,  9, 10, 11],
       [ 1,  5,  6,  7,  8],
       [ 0,  1,  2,  3,  4]])

In [37]:
test = np.array([[1,2,3,4],[2,3,4,5]])
np.flip(test),test.T

(array([[5, 4, 3, 2],
        [4, 3, 2, 1]]),
 array([[1, 2],
        [2, 3],
        [3, 4],
        [4, 5]]))

In [38]:
test = np.array([[1,2,3],[2,1,3],[3,2,1]])
test

array([[1, 2, 3],
       [2, 1, 3],
       [3, 2, 1]])

In [39]:
test.T

array([[1, 2, 3],
       [2, 1, 2],
       [3, 3, 1]])

In [40]:
np.flip(test)

array([[1, 2, 3],
       [3, 1, 2],
       [3, 2, 1]])

In [41]:
def is_countersym_np(A):
    transpose = A.T
    flip = np.flip(A)
    return np.array_equal(flip,transpose)

In [42]:
is_countersym_np(cs_matrix)

True

In [43]:
grader.check("1d")

1d results: All test cases passed!

**1(e)** (5 pts extra credit) [Sudoku](https://en.wikipedia.org/wiki/Sudoku) is a number game played on a $9\times 9$ grid, arranged into nine $3\times 3$ subgrids. In order to win Sudoku, you must fill in the numbers 1-9 exactly once in every row, column, and $3\times 3$ subgrid. Here is an example of a winning solution:

![sudoku](sudoku_solved.png)

*Generalized Sudoku* is played on an $n^2 \times n^2$ grid, arranged into $n^2$ subgrids of size $n\times n$. A winning solution contains the numbers 1-$n$ exactly once in every row, column, and $n\times n$ subgrid.

Write a function that takes a $n^2\times n^2$ integer array and returns `True` if it is a winning (generalized) Sudoku solution.

In [44]:
def sudoku_win_py(A):
    "Returns True if A is a winning Sudoku board"
    import math
    import itertools

    n = math.isqrt(A.shape[0])
    s = set(range(1, n ** 2 + 1))
    horiz, vert = [all(set(row) == s for row in M) for M in (A, A.T)]
    if not (horiz and vert):
        return False
    for h, v in itertools.product(range(0, n * n, n), repeat=2):
        block = A[h : h + n, v : v + n]
        if set(block.flat) != s:
            return False
    return True

In [45]:
good_board = [[3, 1, 6, 2, 4, 9, 7, 8, 5], 
              [2, 7, 8, 3, 6, 5, 9, 1, 4], 
              [4, 9, 5, 7, 8, 1, 6, 2, 3], 
              [8, 5, 1, 9, 7, 4, 3, 6, 2], 
              [9, 2, 3, 8, 1, 6, 4, 5, 7], 
              [6, 4, 7, 5, 3, 2, 1, 9, 8], 
              [7, 6, 2, 1, 5, 3, 8, 4, 9], 
              [1, 3, 9, 4, 2, 8, 5, 7, 6], 
              [5, 8, 4, 6, 9, 7, 2, 3, 1]]
sudoku_win_py(np.array(good_board, dtype=int))

True

In [46]:
bad_board = np.arange(81).reshape(9, 9)
sudoku_win_py(bad_board)

False

In [47]:
np.unique(good_board,axis = 1)

array([[1, 2, 3, 4, 5, 6, 7, 8, 9],
       [7, 3, 2, 6, 4, 8, 9, 1, 5],
       [9, 7, 4, 8, 3, 5, 6, 2, 1],
       [5, 9, 8, 7, 2, 1, 3, 6, 4],
       [2, 8, 9, 1, 7, 3, 4, 5, 6],
       [4, 5, 6, 3, 8, 7, 1, 9, 2],
       [6, 1, 7, 5, 9, 2, 8, 4, 3],
       [3, 4, 1, 2, 6, 9, 5, 7, 8],
       [8, 6, 5, 9, 1, 4, 2, 3, 7]])

In [48]:
good_board

[[3, 1, 6, 2, 4, 9, 7, 8, 5],
 [2, 7, 8, 3, 6, 5, 9, 1, 4],
 [4, 9, 5, 7, 8, 1, 6, 2, 3],
 [8, 5, 1, 9, 7, 4, 3, 6, 2],
 [9, 2, 3, 8, 1, 6, 4, 5, 7],
 [6, 4, 7, 5, 3, 2, 1, 9, 8],
 [7, 6, 2, 1, 5, 3, 8, 4, 9],
 [1, 3, 9, 4, 2, 8, 5, 7, 6],
 [5, 8, 4, 6, 9, 7, 2, 3, 1]]

In [49]:
sudoku_win_np(good_board)

NameError: name 'sudoku_win_np' is not defined

In [50]:
def sudoku_win_np(A):
    A = np.array(A)
    if ((np.unique(A,axis = 1).shape == np.unique(A, axis = 0).shape) and
        (np.unique(A, axis = 0).shape == distinct_rows_np(A).shape)):
        if len(np.unique(A)) < 10:
            return True
    return False

In [51]:
grader.check("1e")

1e results: All test cases passed!

## Question 2: $k$-means clustering

$k$-means is a fundamental algorithm for clustering multivariate data. The inputs to the algorithm are:
- An $n\times p$ data matrix $X$ consisting of $n$ observations of a $p$-dimensional feature vector, and
- A $k\times p$ matrix $C$ containing initial guesses for each $k$ cluster centers.

The algorithm proceeds by iteratively a) assigning each point to the nearest cluster center, and b) recomputing the cluster centers as the mean of all of the currently assigned points. Here is a partial implementation:

In [52]:
def kmeans(X, C):
    """
    K-means algorithm.

    Args:
        - X: ndarray, shape (n, p), n observations of p-dimensional feature vector
        - C: ndarray, shape (k, p), k initial cluster centers

    Returns:
        Tuple of length two:
        The first entry is integer ndarray, shape (n), cluster assignments for each data point
        The second entry is ndarray, shape (k, p), centers of each cluster
    """
    assert X.shape[1] == C.shape[1]  # p should match
    while True:
        new_assignments = nearest_cluster(X, C)
        try:
            
            if np.all(new_assignments == assignments):
                # converged
                return assignments, C
        except NameError:  # first iteration, no assignments
            pass 
        assignments = new_assignments
        C = compute_centroids(X, assignments)

You will finish implementing this algorithm by completing the missing functions `nearest_cluster()` and `compute_centroids()` below. Note: as in the Question 1, do not use any loops, iterators, or comprehensions.

**2(a)** (3 pts) Implement the function `nearest_cluster`. It should take two array arguments, the data points `X` and the cluster centers `C`, and return an integer array giving the index in `C` which is nearest to each point in `X`.

In [53]:
def nearest_cluster(X, C):
    """
    For each point in X, find the nearest point in C.

    Args:
        X: ndarray, shape (n, p), n points of dimension p.
        C: ndarray, shape (k, p), k points of dimension p.

    Returns:
        Integer array of length n, [j[1], j[2], ..., j[n]], such that |X[i] - C[j[i]]| <= |X[i] - C[ell]| for 1 <= ell <= k.
    """
#     assert X.shape[1] == C.shape[1] # X and C should have the same p dimension
#     print(X.shape)
    D = X[:,None,:] - C[None,:,:]
    e = np.linalg.norm(D , axis = 2)
    ans = e.argmin(axis = 1)
    return ans

In [54]:
nearest_cluster(np.array([[-1], [0.5], [1], [1.3], [1.6], [2], [2.3], [2.7], [3], [100]]), np.array([[1], [2], [3]]))

array([0, 0, 0, 0, 1, 1, 1, 2, 2, 2])

In [55]:
grader.check("2a")

2a results: All test cases passed!

**2(b)** (3 pts) Implement the function `compute_centroids`. It should take two array arguments, the data points `X` and the assignment array `a`, and return an $k \times p$ array containing the cluster centroids (averages) for each point assigned to cluster $0, \dots, k-1$. (You may assume that every entry of $a$ is between $0$ and $k-1$, inclusive.)

In [56]:
x = np.linspace(0, 15, 16).reshape(4,4)
x ,np.add.reduceat(x, [0,3,1,2,1])

(array([[ 0.,  1.,  2.,  3.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.],
        [12., 13., 14., 15.]]),
 array([[12., 15., 18., 21.],
        [12., 13., 14., 15.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.],
        [24., 27., 30., 33.]]))

In [57]:
def compute_centroids(X, a): 
    
    combined = np.hstack((X,a[:,None]))
    sorted_combine = combined[combined[:,-1].argsort()]
    
    ele, index, count =  np.unique(sorted_combine[:,-1], return_index = True , return_counts = True)
    theSum = np.add.reduceat(sorted_combine[:,:-1], index, axis = 0)

    theMean = theSum / count[:,None]
    return theMean

In [58]:
X = np.random.rand(30, 2)
a = np.random.randint(0, 5, size=30)

In [1805]:
X

array([[0.76677459, 0.50470061],
       [0.49805562, 0.84289984],
       [0.06780693, 0.57327227],
       [0.94276258, 0.51786005],
       [0.19446581, 0.84793939],
       [0.25163914, 0.70072604],
       [0.54026094, 0.94883629],
       [0.6243367 , 0.83797796],
       [0.00793288, 0.98934006],
       [0.07771457, 0.32212951],
       [0.9461524 , 0.0089391 ],
       [0.82273001, 0.86121164],
       [0.43983081, 0.25574519],
       [0.80268952, 0.477862  ],
       [0.13433861, 0.92784899],
       [0.89596997, 0.49154515],
       [0.8567025 , 0.41857803],
       [0.6834649 , 0.39799064],
       [0.50574202, 0.1895517 ],
       [0.96498892, 0.29421569],
       [0.10345956, 0.1443154 ],
       [0.01409229, 0.7159457 ],
       [0.56449832, 0.79457835],
       [0.50707992, 0.79182104],
       [0.69576421, 0.77784848],
       [0.40648287, 0.64777063],
       [0.1797943 , 0.32181996],
       [0.17260462, 0.40863723],
       [0.24141875, 0.40692199],
       [0.97522234, 0.32031933]])

In [1806]:
np.argsort(a)

array([23, 22,  5,  9, 10, 11, 15, 24,  3,  8, 17, 28,  0, 21, 20, 18, 14,
        7,  4, 16, 29, 19,  6,  1, 26, 27, 13, 12,  2, 25])

In [1807]:
a[:,None]

array([[2],
       [3],
       [4],
       [1],
       [2],
       [0],
       [3],
       [2],
       [1],
       [0],
       [0],
       [0],
       [4],
       [4],
       [2],
       [0],
       [2],
       [1],
       [2],
       [3],
       [2],
       [2],
       [0],
       [0],
       [1],
       [4],
       [3],
       [3],
       [1],
       [2]])

In [1808]:
combined = np.hstack((X,a[:,None]))


In [1809]:
sorted_combine = combined[combined[:,2].argsort()]
sorted_combine

array([[0.50707992, 0.79182104, 0.        ],
       [0.56449832, 0.79457835, 0.        ],
       [0.25163914, 0.70072604, 0.        ],
       [0.07771457, 0.32212951, 0.        ],
       [0.9461524 , 0.0089391 , 0.        ],
       [0.82273001, 0.86121164, 0.        ],
       [0.89596997, 0.49154515, 0.        ],
       [0.69576421, 0.77784848, 1.        ],
       [0.94276258, 0.51786005, 1.        ],
       [0.00793288, 0.98934006, 1.        ],
       [0.6834649 , 0.39799064, 1.        ],
       [0.24141875, 0.40692199, 1.        ],
       [0.76677459, 0.50470061, 2.        ],
       [0.01409229, 0.7159457 , 2.        ],
       [0.10345956, 0.1443154 , 2.        ],
       [0.50574202, 0.1895517 , 2.        ],
       [0.13433861, 0.92784899, 2.        ],
       [0.6243367 , 0.83797796, 2.        ],
       [0.19446581, 0.84793939, 2.        ],
       [0.8567025 , 0.41857803, 2.        ],
       [0.97522234, 0.32031933, 2.        ],
       [0.96498892, 0.29421569, 3.        ],
       [0.

In [1810]:
np.unique(sorted_combine[:,2], return_index = True)[1][1:]

array([ 7, 12, 21, 26])

In [1811]:
ele, index, count =  np.unique(sorted_combine[:,2], return_index = True , return_counts = True)
ele, index, count

(array([0., 1., 2., 3., 4.]),
 array([ 0,  7, 12, 21, 26]),
 array([7, 5, 9, 5, 4]))

In [1812]:
ans = np.split(sorted_combine[:,:2], np.unique(sorted_combine[:,2], return_index = True)[1][1:])

In [1813]:
ans

[array([[0.50707992, 0.79182104],
        [0.56449832, 0.79457835],
        [0.25163914, 0.70072604],
        [0.07771457, 0.32212951],
        [0.9461524 , 0.0089391 ],
        [0.82273001, 0.86121164],
        [0.89596997, 0.49154515]]),
 array([[0.69576421, 0.77784848],
        [0.94276258, 0.51786005],
        [0.00793288, 0.98934006],
        [0.6834649 , 0.39799064],
        [0.24141875, 0.40692199]]),
 array([[0.76677459, 0.50470061],
        [0.01409229, 0.7159457 ],
        [0.10345956, 0.1443154 ],
        [0.50574202, 0.1895517 ],
        [0.13433861, 0.92784899],
        [0.6243367 , 0.83797796],
        [0.19446581, 0.84793939],
        [0.8567025 , 0.41857803],
        [0.97522234, 0.32031933]]),
 array([[0.96498892, 0.29421569],
        [0.54026094, 0.94883629],
        [0.49805562, 0.84289984],
        [0.1797943 , 0.32181996],
        [0.17260462, 0.40863723]]),
 array([[0.80268952, 0.477862  ],
        [0.43983081, 0.25574519],
        [0.06780693, 0.57327227],
      

In [1814]:
np.sum(ans[0], axis = 0)

array([4.06578432, 3.97095083])

In [1815]:
np.add.reduceat(sorted_combine[:,:2],index[0:], axis = 0) 

array([[4.06578432, 3.97095083],
       [2.57134331, 3.08996121],
       [4.17513441, 4.90717711],
       [2.35570441, 2.81640901],
       [1.71681014, 1.9546501 ]])

In [1816]:
compute_centroids(X, a)

array([[0.58082633, 0.56727869],
       [0.51426866, 0.61799224],
       [0.46390382, 0.5452419 ],
       [0.47114088, 0.5632818 ],
       [0.42920253, 0.48866253]])

In [1817]:
test = np.zeros((5,2))
np.average(X[a == 0], axis = 0)

array([0.58082633, 0.56727869])

In [1818]:
np.allclose(X[a == 0].mean(axis=0), compute_centroids(X, a)[0])

True

In [1819]:
X[a == 0].mean(axis=0)

array([0.58082633, 0.56727869])

In [1820]:
inserted = np.insert(test,a,X,axis = 0)
inserted

array([[0.25163914, 0.70072604],
       [0.07771457, 0.32212951],
       [0.9461524 , 0.0089391 ],
       [0.82273001, 0.86121164],
       [0.89596997, 0.49154515],
       [0.56449832, 0.79457835],
       [0.50707992, 0.79182104],
       [0.        , 0.        ],
       [0.94276258, 0.51786005],
       [0.00793288, 0.98934006],
       [0.6834649 , 0.39799064],
       [0.69576421, 0.77784848],
       [0.24141875, 0.40692199],
       [0.        , 0.        ],
       [0.76677459, 0.50470061],
       [0.19446581, 0.84793939],
       [0.6243367 , 0.83797796],
       [0.13433861, 0.92784899],
       [0.8567025 , 0.41857803],
       [0.50574202, 0.1895517 ],
       [0.10345956, 0.1443154 ],
       [0.01409229, 0.7159457 ],
       [0.97522234, 0.32031933],
       [0.        , 0.        ],
       [0.49805562, 0.84289984],
       [0.54026094, 0.94883629],
       [0.96498892, 0.29421569],
       [0.1797943 , 0.32181996],
       [0.17260462, 0.40863723],
       [0.        , 0.        ],
       [0.

In [1821]:
compute_centroids(X,a)

array([[0.58082633, 0.56727869],
       [0.51426866, 0.61799224],
       [0.46390382, 0.5452419 ],
       [0.47114088, 0.5632818 ],
       [0.42920253, 0.48866253]])

In [1822]:
X[a == 0].mean(axis = 0)

array([0.58082633, 0.56727869])

In [1823]:
compute_centroids(X,a)[0]

array([0.58082633, 0.56727869])

In [1824]:
grader.check("2b")

2b results: All test cases passed!

In [1825]:
X = np.random.rand(30, 2)
a = np.random.randint(0, 5, size=30)
test = np.zeros((5,2))

In [1826]:
boardCast = test[:,None,:] +X
np.put_along_axis(test,a[:,None], X,axis = 0)  # it will override the previous result
test

array([[0.05991769, 0.12134346],
       [0.90853515, 0.62336012],
       [0.58575927, 0.96959575],
       [0.55624023, 0.13645523],
       [0.80063267, 0.23297427]])

**2(c)** (5 pts.) The performance of the $k$-means algorithm is known to depend heavily on the starting point (the initial clusters `C` passed in as the second argument.) In some cases, using a "good" starting point can dramatically improve the performance of the algorithm.

The $k$-means++ algorithm is designed to find such a good starting point. [According to Wikipedia](https://en.wikipedia.org/wiki/K-means%2B%2B), the steps of $k$-means++ are:

1. Choose one center uniformly at random among the data points.
2. For each data point $x$ not chosen yet, compute $D(x)$, the distance between $x$ and the nearest center that has already been chosen.
3. Choose one new data point at random as a new center, using a weighted probability distribution where a point $x$ is chosen with probability proportional to $D(x)^2$.
4. Repeat Steps 2 and 3 until $k$ centers have been chosen.

Implement this algorithm using the skeleton provided below. As before, your implementation should only use Numpy functions--no additional loops or comprehensions. 

**Note**: To ensure reproducibility, the parts of the algorithm that rely on ranndomness are provided for you. Your job is to fill in the missing lines necessary to complete the algoritm.

In [114]:
def kmeanspp(X, k, rng):
    """
    k-means++ algorithm.

    Args:
        - X: ndarray, shape (n, p), as above.
        - k, the number of clusters.
        - rng: instance of np.random.Generator().

    Returns:
        ndarray, shape (k, p), cluster centers.
    """
    n, p = X.shape
    C = np.zeros((k, p))
    # step 1
    j = rng.choice(n)
    C[0] = X[j]
    
    for i in range(1, k):
        # step 2
#         deleted = np.delete(X,j , axis = 0)


#         D_x = np.linalg.norm(X - C[i-1],axis = 1)
#         D_x = D_x**2
        dist = np.sum(np.square(X-C[i-1]),axis = 1)
#         n, p = deleted.shape
        # calculate the w
        
#         w = D_x/np.sum(D_x)  # the vectorized probability
        w = dist/np.sum(dist)
        # step 3
        j = rng.choice(n, p=w)
        C[i] = X[j]
    return C

In [115]:
a = np.random.default_rng(12345)
j = a.choice(30)
j

20

In [116]:
X = np.random.rand(30, 4)
k = 2
rng = np.random.default_rng(12345)

In [117]:
n, p = X.shape
C = np.zeros((k, p))
# step 1
j = rng.choice(n)
C,j
C[0] = X[j]

In [118]:
deleted = np.delete(X,j , axis = 0)
D_x = np.linalg.norm(deleted - C[0],axis = 1) 
D_x , deleted.shape , D_x.shape

(array([1.44583065, 0.42455024, 0.98776132, 0.78806801, 0.25062497,
        1.15524309, 1.22595197, 1.12530097, 0.84926042, 0.78444021,
        0.79438924, 1.28766937, 0.77247576, 0.55867606, 0.97928658,
        0.76152552, 1.02907699, 1.03606089, 0.42085579, 1.07898642,
        0.65127131, 0.71156649, 1.18970133, 1.18992199, 1.13723174,
        0.82533201, 0.59117673, 0.52830595, 1.06520319]),
 (29, 4),
 (29,))

In [119]:
D_x = D_x**2
D_x

array([2.09042628, 0.18024291, 0.97567243, 0.62105119, 0.06281287,
       1.33458661, 1.50295822, 1.26630226, 0.72124326, 0.61534644,
       0.63105427, 1.6580924 , 0.5967188 , 0.31211894, 0.9590022 ,
       0.57992111, 1.05899945, 1.07342218, 0.17711959, 1.1642117 ,
       0.42415432, 0.50632687, 1.41538926, 1.41591434, 1.29329603,
       0.68117293, 0.34948992, 0.27910718, 1.13465785])

In [120]:
sum(D_x)

25.080811810510834

In [121]:
dist = np.sqrt(np.sum(np.square(X-C[0]),axis = 1))
dist

array([1.44583065, 0.42455024, 0.98776132, 0.78806801, 0.25062497,
       1.15524309, 1.22595197, 1.12530097, 0.84926042, 0.78444021,
       0.79438924, 1.28766937, 0.77247576, 0.55867606, 0.97928658,
       0.76152552, 1.02907699, 1.03606089, 0.42085579, 1.07898642,
       0.        , 0.65127131, 0.71156649, 1.18970133, 1.18992199,
       1.13723174, 0.82533201, 0.59117673, 0.52830595, 1.06520319])

In [122]:
w = D_x/sum(D_x)

In [123]:
j = rng.choice(29, p = D_x/np.sum(D_x))
j

7

In [124]:
grader.check("2c")

2c results: All test cases passed!

**2(d)** (2 pts) In order to measure how good a clustering is, we can define the *within-class variance* 

$$ V(\mathbf{X}, \mathbf{a}, \mathbf{C}) = \sum_{i=1}^n \| \mathbf{x}_i - \mathbf{c}_{a_i} \|^2,$$

where the $i$-th element of $\mathbf{a}=\{a_1,\dots,a_n\}$ is the cluster assignment of observation $i$, and $\mathbf{C}=(\mathbf{c}_1,\dots,\mathbf{c}_k)$ are the centers of each cluster. Thus, $V(\mathbf{X}, \mathbf{a}, \mathbf{C})$ is the sum of the squared distance from each data point to the center of its assigned cluster.

Implement this function. (Again, no loops, just use Numpy functions.)

In [126]:
def V(X, a, C):
    rows , cols = X.shape
    temp = np.zeros((rows,cols))
    temp = C[a]
    ans = X - temp
    ans = ans**2
    return np.sum(ans)

In [127]:
X = np.arange(9).reshape(3, 3)

a = np.arange(3)

C = 2 + X

C ,X , a

(array([[ 2,  3,  4],
        [ 5,  6,  7],
        [ 8,  9, 10]]),
 array([[0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]]),
 array([0, 1, 2]))

In [128]:
combined = np.hstack((X,a[:,None]))
sorted_combine = combined[combined[:,2].argsort()]

ele, index, count =  np.unique(sorted_combine[:,2], return_index = True , return_counts = True)
theSum = np.add.reduceat(sorted_combine[:,:2], index, axis = 0)

In [129]:
a = np.random.randint(0, 5, size=30)

x = np.random.rand(30, 2)

c = np.random.rand(5,2)

In [130]:
c,a

(array([[0.60431048, 0.54954792],
        [0.92618143, 0.91873344],
        [0.39487561, 0.96326253],
        [0.17395567, 0.12632952],
        [0.13507916, 0.50566217]]),
 array([4, 0, 0, 3, 1, 1, 0, 0, 4, 2, 3, 4, 2, 0, 3, 3, 1, 2, 4, 3, 0, 0,
        4, 2, 4, 2, 0, 3, 0, 0]))

In [131]:
test = np.zeros((30,2))
test = c[a]
ans = np.abs(x - test)
ans = ans**2
ans

array([[8.21382963e-03, 4.29844081e-02],
       [1.98858032e-03, 2.88360345e-01],
       [2.83381833e-01, 1.74497023e-01],
       [1.55350120e-01, 5.92341346e-03],
       [4.54081480e-01, 3.05926621e-02],
       [5.33998406e-01, 1.13821559e-01],
       [1.33743443e-01, 8.83759216e-02],
       [1.32833077e-01, 3.11120840e-03],
       [2.35105279e-01, 1.04535004e-01],
       [5.66840950e-02, 8.92432255e-01],
       [1.08021772e-02, 1.29611226e-01],
       [2.22076849e-01, 3.99288402e-03],
       [6.00829673e-03, 6.42806350e-04],
       [6.03452179e-04, 2.86986466e-02],
       [1.42123385e-01, 3.83167080e-01],
       [2.45299531e-01, 1.92071987e-02],
       [7.39336162e-01, 3.01015986e-01],
       [5.51507152e-02, 5.67142317e-01],
       [3.81524130e-01, 1.92831367e-01],
       [7.45795148e-03, 4.60260541e-01],
       [1.68819252e-01, 8.08434026e-03],
       [6.34255728e-03, 1.40820104e-01],
       [1.64397561e-02, 1.93337035e-01],
       [1.15729474e-01, 3.65132857e-02],
       [5.971219

In [132]:
grader.check("2d")

2d results: All test cases passed!

<!-- BEGIN QUESTION -->

**2(e)** (5 pts) Recall from lecture the file `mnist.npz`, which contains the labeled image data for handwritten digits.

In [2370]:
def kmeans(X, C):
    """
    K-means algorithm.

    Args:
        - X: ndarray, shape (n, p), n observations of p-dimensional feature vector
        - C: ndarray, shape (k, p), k initial cluster centers

    Returns:
        Tuple of length two:
        The first entry is integer ndarray, shape (n), cluster assignments for each data point
        The second entry is ndarray, shape (k, p), centers of each cluster
    """
    assert X.shape[1] == C.shape[1]  # p should match
    while True:
        new_assignments = nearest_cluster(X, C)
        try:
            
            if np.all(new_assignments == assignments):
                # converged
                return assignments, C
        except NameError:  # first iteration, no assignments
            pass 
        assignments = new_assignments
        C = compute_centroids(X, assignments)


In [2371]:
def nearest_cluster(X, C):
    """
    For each point in X, find the nearest point in C.

    Args:
        X: ndarray, shape (n, p), n points of dimension p.
        C: ndarray, shape (k, p), k points of dimension p.

    Returns:
        Integer array of length n, [j[1], j[2], ..., j[n]], such that |X[i] - C[j[i]]| <= |X[i] - C[ell]| for 1 <= ell <= k.
    """
#     assert X.shape[1] == C.shape[1] # X and C should have the same p dimension
#     print(X.shape)
    D = X[:,None,:] - C[None,:,:]
    e = np.linalg.norm(D , axis = 2)
    ans = e.argmin(axis = 1)
    return ans

In [2372]:
def compute_centroids(X, a): 
    
    combined = np.hstack((X,a[:,None]))
    sorted_combine = combined[combined[:,-1].argsort()]
    
    ele, index, count =  np.unique(sorted_combine[:,-1], return_index = True , return_counts = True)
    theSum = np.add.reduceat(sorted_combine[:,:-1], index, axis = 0)

    theMean = theSum / count[:,None]
    return theMean

In [2373]:
mnist = np.load("mnist.npz")
mnist["images"].shape
ans = mnist["labels"][:1000]

We will experiment with clustering these data. For memory and performance reasons, we will only look at the first 1000 images:

In [2374]:
X = mnist["images"][:1000]
X.shape

(1000, 784)

In [2386]:
rng = np.random.default_rng()

In [2387]:
C = kmeanspp(X,10,rng) # the cluster center points

In [2388]:
len(C)

10

In [2389]:
a = nearest_cluster(X,C)

In [2390]:
compute_centroids(X,a)

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [2391]:
pred, centerPoints =kmeans(X,C) 

In [2428]:
pred = np.array(pred, dtype = "uint8")
(pred)%9

array([3, 3, 6, 0, 2, 1, 8, 3, 8, 2, 3, 2, 6, 5, 8, 0, 1, 4, 8, 2, 6, 3,
       2, 0, 8, 4, 6, 3, 0, 2, 8, 4, 5, 2, 3, 2, 5, 3, 6, 5, 8, 4, 8, 8,
       6, 0, 6, 2, 6, 3, 8, 7, 0, 2, 0, 4, 7, 2, 6, 0, 1, 2, 5, 3, 5, 2,
       5, 0, 7, 7, 8, 2, 8, 5, 6, 3, 0, 0, 0, 2, 7, 3, 1, 5, 0, 4, 6, 0,
       7, 2, 5, 0, 6, 5, 4, 3, 0, 4, 6, 0, 8, 0, 8, 0, 8, 0, 5, 3, 7, 8,
       6, 6, 8, 0, 7, 2, 6, 1, 7, 3, 0, 7, 0, 0, 8, 4, 5, 2, 8, 5, 8, 2,
       8, 8, 6, 3, 6, 4, 7, 2, 8, 2, 6, 6, 6, 2, 4, 5, 8, 6, 8, 5, 8, 8,
       6, 5, 7, 3, 6, 1, 8, 4, 6, 2, 6, 5, 2, 0, 0, 3, 6, 1, 0, 2, 8, 2,
       6, 0, 3, 3, 8, 0, 7, 0, 8, 0, 6, 3, 6, 6, 5, 2, 7, 0, 6, 0, 6, 4,
       6, 1, 8, 0, 4, 4, 5, 0, 7, 0, 8, 7, 7, 8, 8, 1, 6, 3, 6, 2, 5, 3,
       1, 5, 6, 0, 4, 4, 6, 0, 6, 7, 0, 8, 6, 1, 7, 3, 4, 2, 5, 2, 8, 5,
       3, 0, 6, 4, 7, 8, 8, 7, 6, 8, 0, 1, 6, 3, 0, 2, 8, 2, 2, 5, 2, 0,
       6, 4, 6, 8, 8, 0, 8, 2, 0, 2, 6, 2, 8, 1, 5, 4, 2, 3, 6, 7, 6, 8,
       5, 2, 0, 2, 0, 3, 8, 7, 1, 4, 7, 2, 6, 5, 0,

In [2429]:
ans

array([5, 0, 4, 1, 9, 2, 1, 3, 1, 4, 3, 5, 3, 6, 1, 7, 2, 8, 6, 9, 4, 0,
       9, 1, 1, 2, 4, 3, 2, 7, 3, 8, 6, 9, 0, 5, 6, 0, 7, 6, 1, 8, 7, 9,
       3, 9, 8, 5, 9, 3, 3, 0, 7, 4, 9, 8, 0, 9, 4, 1, 4, 4, 6, 0, 4, 5,
       6, 1, 0, 0, 1, 7, 1, 6, 3, 0, 2, 1, 1, 7, 9, 0, 2, 6, 7, 8, 3, 9,
       0, 4, 6, 7, 4, 6, 8, 0, 7, 8, 3, 1, 5, 7, 1, 7, 1, 1, 6, 3, 0, 2,
       9, 3, 1, 1, 0, 4, 9, 2, 0, 0, 2, 0, 2, 7, 1, 8, 6, 4, 1, 6, 3, 4,
       5, 9, 1, 3, 3, 8, 5, 4, 7, 7, 4, 2, 8, 5, 8, 6, 7, 3, 4, 6, 1, 9,
       9, 6, 0, 3, 7, 2, 8, 2, 9, 4, 4, 6, 4, 9, 7, 0, 9, 2, 9, 5, 1, 5,
       9, 1, 2, 3, 2, 3, 5, 9, 1, 7, 6, 2, 8, 2, 2, 5, 0, 7, 4, 9, 7, 8,
       3, 2, 1, 1, 8, 3, 6, 1, 0, 3, 1, 0, 0, 1, 7, 2, 7, 3, 0, 4, 6, 5,
       2, 6, 4, 7, 1, 8, 9, 9, 3, 0, 7, 1, 0, 2, 0, 3, 5, 4, 6, 5, 8, 6,
       3, 7, 5, 8, 0, 9, 1, 0, 3, 1, 2, 2, 3, 3, 6, 4, 7, 5, 0, 6, 2, 7,
       9, 8, 5, 9, 2, 1, 1, 4, 4, 5, 6, 4, 1, 2, 5, 3, 9, 3, 9, 0, 5, 9,
       6, 5, 7, 4, 1, 3, 4, 0, 4, 8, 0, 4, 3, 6, 8,

In [2393]:
kpp_v = V(X, pred, centerPoints)
print("the evaulation of the kpp:",kpp_v )

the evaulation of the kpp: 2419297708.8340654


In [2394]:
### select the random center points

In [2417]:
compare_list = []
for i in range(20):
    ## kpp
    rng = np.random.default_rng()
    C = kmeanspp(X,10,rng)
    a = nearest_cluster(X,C)
    pred, centerPoints =kmeans(X,C) 
    kpp_v = V(X, pred, centerPoints)
    
    ##random 
    
    index = np.random.randint(0, 999, size=10)
    C = X[index]
    pred, centerPoints =kmeans(X,C) 
    pred = np.array(pred, dtype = "uint8")
    random_v = V(X, pred, centerPoints)
    
    
    if random_v > kpp_v:  # if the 
        compare_list.append(1)
    compare_list.append(0)

In [2418]:
np.count_nonzero(compare_list)

13

Which performs better, $k$-means++ or random initialization? Do the clusters make sense to you? How do the clusters relate to the true labels given in `mnist['labels']`? What are some examples of images where the clustering is nearly ambiguous (meaning they were almost part of another cluster?)

_Type your answer here, replacing this text._

Based on the observation, the k-means++ perform slightly better than the random initialization. The above for loop is to compare the sum of the squared distance from each data point to the center of its assigned cluster. It shows the most of the time the k-means++ has the less distance compared to the random initialization. 

We do not know which cluster is associated with which true labels given in the dataset. However based on the distribution of the prediciton, we can say it matches the distribution of the true label. 

The exmaples of the images where the clustering is nearly ambiguous shows the boundary where the distance from the data point to the other center of the cluster is the same. It shows the probabiilty of the two classes is the same.

<!-- END QUESTION -->

## Question 3: Working with pandas DataFrames

In this problem, you'll get practice working with pandas `DataFrames`, reading
them into and out of memory, changing their contents and performing
aggregation operations. We'll use the file `iris.csv` included with this problem set to practice.
**Note:** for the sake of consistency, please the CSV included with the problem set, and not one from elsewhere.

In [64]:
import pandas as pd

<!-- BEGIN QUESTION -->

**3(a)** (2 pts) Read the data into a variable called `iris`. How many data points are
    there in this data set? What are the data types of the columns? What
    are the column names? The column names correspond to flower species
    names, as well as four basic measurements one can make of a flower:
    the width and length of its petals and the width and length of its
    sepal (the part of the pant that supports and protects the flower
    itself). How many species of flower are included in the data? Show your work by including the
    pandas commands you used to figure out the answers.

In [135]:
iris = pd.read_csv("iris.csv")

In [136]:
len(iris)

150

In [137]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  150 non-null    float64
 1   Sepal.Width   150 non-null    float64
 2   Petal.Length  150 non-null    float64
 3   Petal.Width   150 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [92]:
iris.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [93]:
iris.columns

Index(['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species'],
      dtype='object')

In [94]:
np.unique(iris["Species"])

array(['setosa', 'versicolor', 'virginica'], dtype=object)

there are 150 data points
the datatype in the column are float64 and the string
column name:'Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species'

number of species :'setosa', 'versicolor', 'virginica'

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**3(b)** It is now known that this dataset contains errors
    in two of its rows (see the documentation at
    <https://archive.ics.uci.edu/ml/datasets/Iris>). Using 1-indexing,
    these errors are in the 35th and 38th rows. The 35th row should read
    `4.9,3.1,1.5,0.2,"setosa"`, 
    where the fourth feature is incorrect as it appears in the file,
    and the 38th row should read `4.9,3.6,1.4,0.1,"setosa"`, where the second and third features
    are incorrect as they appear in the file. Correct these entries of
    your DataFrame.


In [96]:
iris.iloc[34]

Sepal.Length       4.9
Sepal.Width        3.1
Petal.Length       1.5
Petal.Width        0.2
Species         setosa
Name: 34, dtype: object

In [101]:
iris.iloc[37]

Sepal.Length         4.9
Sepal.Width          3.6
Petal.Length         1.4
Petal.Width          0.1
Species           setosa
Petal.Ratio         14.0
Sepal.Ratio     1.361111
Name: 37, dtype: object

In [109]:
# the dataset is already modified and fixed the errors. We do not need to do anything on this dataset

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**3(c)** The iris dataset is commonly used in machine learning as a
        proving ground for clustering and classification algorithms.
        Some researchers have found it useful to use two additional features,
        called *Petal ratio* and *Sepal ratio*,
        defined as the ratio of the petal length to petal width
        and the ratio of the sepal length to sepal width, respectively.
        Add two columns to your DataFrame corresponding to these two
        new features.
        Name these columns
        `Petal.Ratio` and `Sepal.Ratio`, respectively.

In [138]:
iris["Petal.Ratio"] = iris["Petal.Length"] / iris["Petal.Width"]
iris["Sepal.Ratio"] = iris["Sepal.Length"] / iris["Sepal.Width"]

In [139]:
iris.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species  Petal.Ratio  \
0           5.1          3.5           1.4          0.2  setosa          7.0   
1           4.9          3.0           1.4          0.2  setosa          7.0   
2           4.7          3.2           1.3          0.2  setosa          6.5   
3           4.6          3.1           1.5          0.2  setosa          7.5   
4           5.0          3.6           1.4          0.2  setosa          7.0   

   Sepal.Ratio  
0     1.457143  
1     1.633333  
2     1.468750  
3     1.483871  
4     1.388889

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

**3(d)** (2 pts)
Use a pandas aggregate operation to determine the
        mean, median, minimum, maximum and standard deviation of the
        petal and sepal ratio for each of the three species in the data set.
        **Note**: you should be able to get all five numbers in a single
        table (indeed, in a single line of code)
        using a well-chosen group-by or aggregate operation.

In [140]:
iris.groupby("Species").agg(['mean','median','max','min','std'])

Sepal.Length                            Sepal.Width              \
                   mean median  max  min       std        mean median  max   
Species                                                                      
setosa            5.006    5.0  5.8  4.3  0.352490       3.428    3.4  4.4   
versicolor        5.936    5.9  7.0  4.9  0.516171       2.770    2.8  3.4   
virginica         6.588    6.5  7.9  4.9  0.635880       2.974    3.0  3.8   

                           ... Petal.Ratio                            \
            min       std  ...        mean    median   max       min   
Species                    ...                                         
setosa      2.3  0.379064  ...    6.908000  7.000000  15.0  2.666667   
versicolor  2.0  0.313798  ...    3.242837  3.240385   4.1  2.666667   
virginica   2.2  0.322497  ...    2.780662  2.666667   4.0  2.125000   

                     Sepal.Ratio                                          
                 std        mean    median       max       min       std  
Species                                                                   
setosa      2.854545    1.470188  1.463063  1.956522  1.268293  0.118750  
versicolor  0.312456    2.160402  2.161290  2.818182  1.764706  0.228658  
virginica   0.407367    2.230453  2.169540  2.961538  1.823529  0.246992  

[3 rows x 30 columns]

<!-- END QUESTION -->



---

To double-check your work, the cell below will rerun all of the autograder tests.

In [143]:
grader.check_all()

1a results: All test cases passed!

1b results: All test cases passed!

1c results: All test cases passed!

1d results: All test cases passed!

1e results: All test cases passed!

2a results: All test cases passed!

2b results: All test cases passed!

2c results: All test cases passed!

2d results: All test cases passed!

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

Upload this .zip file to Gradescope for grading.

In [144]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False)